.travis.yml

```diff
@@ -12,18 +12,23 @@ addons:
       - qt5-default
       - qtdeclarative5-dev
       - libqt5charts5-dev
+      - clang-format-5.0
 
 # https://docs.platformio.org/en/latest/integration/ci/travis.html#travis-ci
 cache:
     directories:
         - "~/.platformio"
+        - $HOME/.cache/pre-commit
 
 install:
   - pip install -U platformio
   - platformio update
+  - pip install pre-commit
 
 # TODO: Separate jobs for GUI and controller?
 script:
+  # Run pre-commit for style checks etc.
+  - pre-commit run --all-files --show-diff-on-failure
```

Comment:
...that said I don't think this is going to do what we want for clang-format unfortunately.

The clang-format presubmit checks that the lines *that you modified* are formatted correctly.  But when we run here, we don't know "what you modified".  So I think clang-format will always pass here.

That's probably OK?  I feel like the format check maybe shouldn't block Travis.  If we do want it to run in Travis, I see two options:

1. Check that *all lines* are formatted correctly.  In my experience this can be kind of harsh and can result in spurious failures "that aren't your fault".
2. Check that the lines *from this commit* are formatted correctly.  To do this we'd need to somehow communicate to pre-commit what changed in this push.  You might could do something like

   ```
   git revert origin/master
   pre-commit run --show-diff-on-failure
  ```

to uncommit the changes in this commit *but leave them in the working tree* (I think that's `git revert`?) and then run pre-commit over that diff.

That might work for pre-commit Travis run.  Probably won't work for postcommit Travis runs.  Maybe Travis has a concept of "base commit"?

Anyway that's all more work, if you want we can just leave it as-is with a TODO that this doesn't do the right thing for clang-format, maybe linking back to this comment or smth.